vm/devices/hyperv_ic/src/shutdown.rs

```diff
@@ -137,40 +155,53 @@ impl ShutdownChannel {
+                    };
+                    if self.pending_shutdown.is_some() {
+                        panic!(
+                            "Shutdown request received while processing previous shutdown request"
+                        );
+                    }
+                    self.pending_shutdown = Some(rpc.1);
+                    *state = ReadyState::SendShutdown(rpc.0);
+                }
                 Event::StateMachine(r) => {
                     r?;
                 }
-                Event::Request(req) => match req {
-                    ShutdownRpc::WaitReady(rpc) => match self.state {
-                        ChannelState::SendVersion | ChannelState::WaitVersion => {
-                            ic.wait_ready.push(rpc)
-                        }
-                        ChannelState::Ready { .. } => rpc.complete(()),
-                    },
-                    ShutdownRpc::Shutdown(rpc) => match self.state {
-                        ChannelState::SendVersion | ChannelState::WaitVersion => {
-                            rpc.complete(ShutdownResult::NotReady)
-                        }
-                        ChannelState::Ready { ref mut state, .. } => match state {
-                            ReadyState::Ready => {
-                                self.pending_shutdown = Some(rpc.1);
-                                *state = ReadyState::SendShutdown(rpc.0);
-                            }
-                            ReadyState::SendShutdown { .. } | ReadyState::WaitShutdown => {
-                                rpc.complete(ShutdownResult::AlreadyInProgress)
-                            }
-                        },
-                    },
+                Event::Request(ShutdownRpc::WaitReady(rpc)) => match &mut self.state {
+                    ChannelState::SendVersion | ChannelState::WaitVersion => {
+                        ic.wait_ready.push(rpc)
+                    }
+                    ChannelState::Ready {
+                        ref mut clients,
+                        shutdown_notification,
+                        ..
+                    } => {
+                        let send_shutdown_notification = &shutdown_notification.0;
+                        let (send, recv) = mesh::oneshot();
+                        clients.push(send);
+                        rpc.complete((send_shutdown_notification.clone(), recv));
+                    }
                 },
             }
         }
     }
 
     async fn process_state_machine(
         &mut self,
-        wait_ready: &mut Vec<Rpc<(), ()>>,
-    ) -> Result<(), Error> {
-        match self.state {
+        wait_ready: &mut Vec<
+            Rpc<
+                (),
+                (
+                    mesh::MpscSender<Rpc<ShutdownParams, ShutdownResult>>,
+                    mesh::OneshotReceiver<()>,
+                ),
+            >,
+        >,
+    ) -> Result<Option<Rpc<ShutdownParams, ShutdownResult>>, Error> {
+        match &mut self.state {
```

Comment:
nit: I'd stick with the `match self` so that you don't have the extra derefs below. You just need to add `ref mut` to `shutdown_notification` in the match.